desc/protoparse/linker_test.go

```diff
@@ -906,6 +906,84 @@ func TestLinkerValidation(t *testing.T) {
 			},
 			"foo.proto:4:26: field foobar: option json_name is not allowed on extensions",
 		},
+		{
+			map[string]string{
+				"foo.proto": "syntax = \"proto3\";\n" +
+					"package foo.foo;\n" +
+					"import \"other.proto\";\n" +
+					"service Foo { rpc Bar (Baz) returns (Baz); }\n" +
+					"message Baz {\n" +
+					"  foo.Foo.Bar f = 1;\n" +
```

Comment:
This is surprising - shouldn't this actually resolve to The `foo.Foo.Bar` message defined in `other.proto`? The method should only be able to be referred as the full name(`foo.foo.Foo.Bar` or `.foo.foo.Foo.Bar`) or the local nested name (`Foo.Bar`), right?

I might be unfamiliar with the significance of the individual package elements and whether or not we can import a type based on the parent namespace (the first element in `foo.foo`), so this will help me understand the resolution logic here.

Reply:
I added the test thinking that it would indeed have resolved to the message in the other proto. But the reason it does not is because it does not start with a dot: though it appears to be the fully-qualified name of the message in the other file, without a dot it is interpreted as a partially qualified name -- a relative reference.

So the scoping rules work more-or-less like so:
1. We start off in the scope of `foo.foo.Baz`. So we first look for a composite element named `foo.foo.Baz.foo` (using just the first part of the reference). There is no match.
2. We move to the parent scope, which is package `foo.foo`. So now we look for a composite element named `foo.foo.foo`. Again, no dice.
3. We move to the next scope, which is package `foo` (yes, the components in a package name form a hierarchy -- layers of namespacing that get peeled back one at a time). Now we look for a composite element named `foo.foo`. Success! We've found the package `foo.foo`.
4. We now search in that composite element for the full name: `foo.foo.Foo.Bar`. Success again! We've resolved the element to the method with this fully-qualified name.

FWIW, this particular test case passed even before the actual logic fix in this PR. I'm just "back-filling" some tests related to this stuff, using the output of `protoc` to confirm my knowledge of the process. The output from `protoc` for such a case is the following (which, IMHO, is much less clear than the error message from protoparse):
```
foo.proto:6:3: "foo.Foo.Bar" is not a type.
```

Reply:
Great explanation, that cleared it up for me! My expectation of going from innermost scope and bubbling up the stack of parents was accurate, but I wasn't aware of how the package hierarchy is involved in all this.